src/Lumi/Components/Form.purs

```diff
@@ -129,37 +138,175 @@ build editor = makeStateless (createComponent "Form") render where
+    , validated
+    , form: renderer forest
+    }
 
-     in element (unsafeCreateDOMComponent "lumi-form")
-          { "class": String.joinWith " " $ fold
-                       [ guard inlineTable ["inline-table"]
-                       , guard readonly ["readonly"]
-                       ]
-          , children: surround fieldDivider (map toRow forest)
+
+-- | The default Lumi implementation for rendering a forest of JSX
+-- | form fields.
+defaultRenderForm
+  :: { forceTopLabels :: Boolean
+     , readonly :: Boolean
+     , inlineTable :: Boolean
+     }
+  -> Forest JSX
+  -> JSX
+defaultRenderForm { inlineTable, readonly, forceTopLabels } forest =
+  element (R.unsafeCreateDOMComponent "lumi-form")
+    { class:
+        String.joinWith " " $ fold
+          [ guard inlineTable ["inline-table"]
+          , guard readonly ["readonly"]
+          ]
+    , children:
+        surround fieldDivider (map toRow forest)
+    }
+  where
+    fieldDivider = R.hr { className: "lumi field-divider" }
+
+    toRow = case _ of
+      Child { key, child } ->
+        maybe identity keyed key $ child
+      Wrapper { key, children } ->
+        R.div
+          { key: fromMaybe "" key
+          , children: [ intercalate fieldDivider (map toRow children) ]
+          }
+      Node { label, key, required, validationError, children } ->
+        maybe identity keyed key $ labeledField
+          { label: text body
+              { children = [ label ]
+              , className = toNullable (pure "field-label")
+              }
+          , value: intercalate fieldDivider (map toRow children)
+          , validationError
+          , required
+          , forceTopLabel: forceTopLabels
+          , style: R.css {}
           }
 
+
+-- | Consume `useForm` as a render-prop component. Useful when `useForm`
+-- | would be preferred but you don't want to migrate an entire component
+-- | to React's hooks API.
+-- |
+-- | _Note_: this function should be fully applied, to avoid remounting
+-- | the component on each render.
+formState
+  :: forall props unvalidated result xs
+   . RowToList unvalidated xs
+  => HMap ModifyValidated { | unvalidated } { | unvalidated }
+  => FormBuilder
+       { initialState :: { | unvalidated }
+       , readonly :: Boolean
+       , inlineTable :: Boolean
+       , forceTopLabels :: Boolean
+       | props
+       }
+       { | unvalidated }
+       result
+  -> Hooks.ReactComponent
+      { initialState :: { | unvalidated }
+      , render
+          :: { formData :: { | unvalidated }
+             , setFormData :: ({ | unvalidated } -> { | unvalidated }) -> Effect Unit
+             , setModified :: Effect Unit
+             , reset :: Effect Unit
+             , validated :: Maybe result
+             , form :: JSX
+             }
+          -> JSX
+      | props
+      }
+formState editor = unsafePerformEffect do
+  Hooks.component "FormState" \props -> Hooks.do
+    state <- useForm editor (unsafeCoerce props)
```

Comment:
This use of `unsafeCoerce` is unsafe because `useForm`'s 2nd argument is
```purescript
{ initialState :: { | unvalidated }
, readonly :: Boolean
, inlineTable :: Boolean
, forceTopLabels :: Boolean
| props
}
```

but `props`, in this case, doesn't contain `readonly`, `inlineTable` and `forceTopLabels`. Also, I think we should add a type annotation here to prevent this from typechecking if `useForm` changes its signature, like this:
```purescript
formState editor = unsafePerformEffect do
  Hooks.component "FormState" \props -> Hooks.do
    let props' ::
          { initialState :: { | unvalidated }
          , readonly :: Boolean
          , inlineTable :: Boolean
          , forceTopLabels :: Boolean
          | props
          }
        props' = unsafeCoerce props
    state <- useForm editor props'
    pure (props.render state)
```

Reply:
I got it working with this implementation, but the type is a little different. Did I miss something?